python/dgl/dataloading/neighbor.py

```diff
@@ -5,7 +5,55 @@
 from .. import backend as F
 from ..base import ETYPE
 
-class MultiLayerNeighborSampler(BlockSampler):
+class NeighborSamplingMixin(object):
+    """Mixin object containing common optimizing routines that caches fanout and probability
+    arrays.
+
+    The mixin requires the object to have the following attributes:
+
+    - :attr:`prob`: The edge feature name that stores the (unnormalized) probability.
+    - :attr:`fanouts`: The list of fanouts (either an integer or a dictionary of edge
+      types and integers).
+
+    The mixin will generate the following attributes:
+
+    - :attr:`prob_arrays`: List of DGL NDArrays containing the unnormalized probabilities
+      for every edge type.
+    - :attr:`fanout_arrays`: List of DGL NDArrays containing the fanouts for every edge
+      type at every later.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)       # forward to base classes
+        self.fanout_arrays = []
+        self.prob_arrays = None
+
+    def _build_prob_arrays(self, g):
+        if self.prob is not None:
+            self.prob_arrays = [F.to_dgl_nd(g.edges[etype].data[self.prob]) for etype in g.etypes]
+        elif self.prob_arrays is None:
+            # build prob_arrays only once
+            self.prob_arrays = [nd.array([], ctx=nd.cpu())] * len(g.etypes)
+
+    def _build_fanout(self, block_id, g):
+        assert not self.fanouts is None, \
+            "_build_fanout() should only be called when fanouts is not None"
+        # build fanout_arrays only once for each layer
+        while block_id >= len(self.fanout_arrays):
+            for i in range(len(self.fanouts)):
```

Comment:
L41 - L42 is a bit weird. Is it equivalent to use the minimum of `block_id` and `len(self.fanouts)`?

Reply:
I think it can be rewritten to `for i in range(len(self.fanouts))`.  @nv-dlasalle any idea why it's written this way?

Reply:
I don't remember exactly it was written this way (it also doesn't appear entirely correct). Since we never partially build the fanout arrays, it works as is, but in theory if we built it partially, this wouldn't function properly. Especially now that it's part of a separate class, it probably should be something like to only build the  missing layers.
```
if len(self.fanouts) > len(self.fanout_arrays):
  for i in range(len(self.fanout_arrays), len(self.fanouts)):
```

Reply:
Will fix in the next PR.